src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/DeviceAuthManager.cs

```diff
@@ -0,0 +1,146 @@
+
+        public bool TryCreateDeviceAuthChallengeResponseAsync(HttpResponseHeaders responseHeaders, Uri endpointUri, out string responseHeader)
+        {
+            responseHeader = string.Empty;
+            X509Certificate2 certificate = null;
+
+            if (!DeviceAuthHelper.IsDeviceAuthChallenge(responseHeaders))
+            {
+                return false;
+            }
+            if (!DeviceAuthHelper.CanOSPerformPKeyAuth())
+            {
+                responseHeader = DeviceAuthHelper.GetBypassChallengeResponse(responseHeaders);
+                return true;
+            }
+
+            IDictionary<string, string> challengeData = DeviceAuthHelper.ParseChallengeData(responseHeaders);
+
+            if (!challengeData.TryGetValue("SubmitUrl", out string submitUrl))
+            {
+                submitUrl = endpointUri.AbsoluteUri;
+            }
+
+            try
+            {
+                certificate = FindCertificate(challengeData);
+            }
+            catch (MsalException ex)
+            {
+                if (ex.ErrorCode == MsalError.DeviceCertificateNotFound)
+                {
+                    responseHeader = DeviceAuthHelper.GetBypassChallengeResponse(responseHeaders);
+                    return true;
+                }
+            }
+
+            DeviceAuthJWTResponse responseJWT = GetDeviceAuthJwtResponse(submitUrl, challengeData["nonce"], certificate);
+
+            byte[] signedResponse = SignWithCertificate(responseJWT, certificate);
+
+            FormatResponseHeader(responseJWT, signedResponse, challengeData, out responseHeader);
+
+            return true;
+        }
+
+        private void FormatResponseHeader(
+            DeviceAuthJWTResponse responseJWT,
+            byte[] signedResponse,
+            IDictionary<string, string> challengeData,
+            out string responseHeader)
+        {
+            string signedJwt = $"{responseJWT.GetResponseToSign()}.{Base64UrlHelpers.Encode(signedResponse)}";
+            string authToken = $"AuthToken=\"{signedJwt}\"";
+
+            responseHeader = $"PKeyAuth {authToken}, Context=\"{challengeData["Context"]}\", Version=\"{challengeData["Version"]}\"";
+        }
+
+        private X509Certificate2 FindCertificate(IDictionary<string, string> challengeData)
+        {
+            var store = new X509Store(StoreName.My, StoreLocation.CurrentUser);
+            try
+            {
+                store.Open(OpenFlags.ReadOnly);
+                var certCollection = store.Certificates;
+                if (challengeData.ContainsKey("CertAuthorities"))
+                {
+                    return FindCertificateByCertAuthorities(challengeData, certCollection);
+                }
+
+                X509Certificate2Collection signingCert = null;
+                signingCert = certCollection.Find(X509FindType.FindByThumbprint, challengeData["CertThumbprint"],
+                    false);
+                if (signingCert.Count == 0)
```

Comment:
can `signingCert` be null?

Reply:
So, it can't be null here because Find returns at least empty collection. Source for [X509Certificate2Collection.Find](https://referencesource.microsoft.com/#System/security/system/security/cryptography/x509/x509certificate2collection.cs,174) and [X509Utils.GetCertificates](https://referencesource.microsoft.com/#System/security/system/security/cryptography/x509/x509utils.cs,522). But good point, had to look it up. :)

Reply:
okay. thanks for checking. and +1 to combining those lines, if you don't mind.